_cite/plugins/orcid.py

```diff
@@ -35,25 +36,34 @@ def query(_id):
     # go through response structure and pull out ids e.g. doi:1234/56789
     for work in response:
         # get list of ids
-        ids = get_safe(work, "external-ids.external-id", [])
+        ids = []
         for summary in get_safe(work, "work-summary", []):
             ids = ids + get_safe(summary, "external-ids.external-id", [])
+        ids = ids + get_safe(work, "external-ids.external-id", [])
 
-        # prefer doi id type, or fallback to first id
+        # prefer particular "relationship" type, or fallback to first id
         _id = next(
-            (id for id in ids if get_safe(id, "external-id-type", "") == "doi"),
-            ids[0] if len(ids) > 0 else {},
+            (
+                id
+                for id in ids
+                if get_safe(id, "external-id-relationship", "")
+                in ["self", "version-of", "part-of"]
```

Comment:
Based on https://support.orcid.org/hc/en-us/articles/360006894774-Group-multiple-versions-of-the-same-work-together I think that you should always choose "self" over "version-of"  or "part-of". I'm unsure if the list is always sorted with "self" coming before the others. In the few examples I could find (e.g., [Casey's publication history](https://pub.orcid.org/v3.0/0000-0001-8713-9213/works)) that seemed to be the case, in which case you don't need to change this.

To be sure, though, you might consider sorting `ids` according to the relationship priority, e.g.:

```python
# little helper used for sorting
# gets the position of "v" in iterable "l" or puts it at the end if not found
def index_or(l, v):
  return l.index(v) if v in l else len(l)
  
# to get ids sorted in the order specified in the list, with anything not recognized coming after
sorted(ids, key=lambda x: index_or(["self", "part-of"], get_safe(id, "external-id-relationship", "")))

# you'd use that sorted() expression above as the thing you're iterating over
```

It's not the world's most efficient code, but it should be fine for the small number of elements you're dealing with.

Reply:
For Casey's ORCID I'm actually not seeing any ID that's not "self".

<details><summary>script</summary>

```js
const response = await (
  await fetch("https://pub.orcid.org/v3.0/0000-0001-8713-9213/works", {
    headers: { Accept: "application/json" },
  })
).json();
new Set(
  response.group
    .map((g) =>
      g["external-ids"]["external-id"].map(
        (id) => id["external-id-relationship"]
      )
    )
    .flat()
);
```

</details>

I feel like maybe we should keep the "order is the priority" approach here that we're now doing for the work summaries, even though I don't see any documentation to support it.

[This doc page](https://support.orcid.org/hc/en-us/articles/360006971353-Metadata-in-the-Works-section?input_string=api+help+with+%2Fworks+endpoint) explains a bit more what the relationship types could be. Notably, preprints can show up as "version of", which is why I wanted to include it as well. Conceivably, an author could specify that the preprint is the preferred source (probably temporarily) and it might show up first in the list.

Again, wish I had someone from ORCID to consult with. Still haven't been approved access to that google group for help. A lot of this could come down to user preference as well, so keeping it easy to tweak the script is also a priority here.

Reply:
Okay that settles it for me, for now. Let's keep it as is. Can always revert the behavior in a future version... and in lieu of any ORCID authority, seems like the only way to truly find out if it works is to just try it out with many users and see what happens.

Reply:
@falquaddoomi Regarding your idea of having the plugin here match what the ORCID website UI does. I think that's a good idea, though for this part of the code in particular, I don't see anything in their UI that matches this data. They only seem to have a [search](https://orcid.org/orcid-search/search) for authors, not individual papers. And looking at a particular [author page](https://orcid.org/0000-0003-4297-8747), I don't see any UI elements with an order other than the work-summary user-specified order that we've been discussing.

Also related... I'm going to remove the line of code that includes the top level `external-ids` after the `work-summary` `external-ids`. Testing the following script with Casey's, Andrew's, and Faisal's ORCIDs, the top level ids are always a subset of the work summary ids (sometimes the latter has one or two extra). I guess they just duplicate the ids at the top level for convenience (though the fact that they're not exactly the same as the work summary ids is bound to cause issues).

<details><summary>script</summary>

```js
const result = await (
  await fetch("https://pub.orcid.org/v3.0/0000-0001-8713-9213/works", {
    headers: { Accept: "application/json" },
  })
).json();

for (const entry of result.group) {
  const ids = new Set(
    entry["external-ids"]["external-id"].map((id) => id["external-id-value"])
  );
  const summaries = new Set(
    entry["work-summary"]
      .map((summary) =>
        summary["external-ids"]["external-id"].map(
          (id) => id["external-id-value"]
        )
      )
      .flat()
  );
  if (!ids.isSubsetOf(summaries)) console.log({ ids, summaries });
}
```

</details>

IN CONCLUSION... what we're left with is dead simple and I think the most correct. Go through work summaries in order (user preference), and take the first available one (as long as it has an acceptable relationship type). I thought about adding code to sort manubot-citable ID types to the front of the list (in case the first entry is a [weird/obscure type](https://github.com/manubot/manubot/issues/365)). But then I thought it's probably better to respect the user-selected preference. They may have a good reason for putting that weird ID first. The plugin will keep the ORCID citation details anyway.

Hopefully this settles the matter.